rfcs/040-tei_adapter/README.md

```diff
@@ -0,0 +1,108 @@
+It's used at Wellcome to describe manuscripts in non-Latin languages.
+The TEI files are in a [GitHUb Repository](https://github.com/wellcomecollection/wellcome-collection-tei). At the 
+time of writing in April 2021 there are 622 TEI files, but the collection of manuscripts is much larger, around 10k.
+
+Currently, all information stored in TEI files is not surfaced anywhere except the GitHub repository. 
+Many of the TEI files have a Sierra record that they link to, but not all of them – and the record contains minimal information. 
+## How we do it
+### Do we need an adapter?
+
+We need to be able to propagate updates to TEI files as they happen through to the API, and we also need to be able to
+reindex when we make changes in the pipeline. 
+
+GitHub is a relatively reliable and low latency source, so we could in theory just send the changes to a TEI transformer 
+and do a reindex by cloning the repository and/or parsing the tree of the TEI files. 
+There are some disadvantages to this approach:
+- We need a "special" reindexer for TEI.
+- We can't keep track of deleted files. If something is deleted, the change will go through the pipeline once, 
+  but after that there's no way of "remembering" that the record ever existed.
+- We need to assign versions to works resulting from TEI files in order to ensure subsequent updates are 
+  processed correctly in the pipeline. This versioning logic would have to be added explicitly to the 
+  TEI transformer whereas having an adapter and an intermediate store would allow us to use the already existing versioning
+  logic in the versioned store library.
+- Currently, we compare the number of record in the source table to the number of works in the transformer pipeline store
+  to track the success of a reindex. We wouldn't be able to do that without a store.
+
+For these reasons the proposal is to have a TEI adapter that listens to changes to TEI files and a TEI store.
+
+However, since GitHub is a relatively reliable and low latency source, we  could only store the HTTP URLs
+to the TEI files in a TEI store, and not the full XML. This has the advantage that we don't have to manage an 
+ever-growing S3 bucket (we don't have a way of cleaning up old versions of a record), but it has some disadvantages:
+- GitHub outages (which are rare but happen) might affect our reindexes.
+- We might be throttled during a reindex
+- We go to the public internet for every TEI file for every reindex as opposed to getting each file exactly once 
+  per every change that affects it.
+  
+For these reasons, the proposal here is to store the whole TEI file in a full VHS.
+### Store structure
+
+Therefore, taking as an example
+[WMS_Arabic_529](https://github.com/wellcomecollection/wellcome-collection-tei/blob/master/Arabic/WMS_Arabic_529.xml), 
+the proposed store record would be:
+
+| Id | Payload | deleted  | version |
+| --- | ------- | -------| --------|
+| WMS_Arabic_529 | [pointer to location in s3] | false | 1 |
```

Comment:
Just a question on this ID, and around moving of items.

It seems like the XML does have an ID: `manuscript_16172` - would it not be better to use that so that if they change the filenames, we don't delete and create new works risking losing access to URLs?

Reply:
I picked what is in `<idno type="msID">WMS_Arabic_529</idno>` because @adrianplau said that's the id used for TEI (if I understood correctly)>. I'm not actually sure what that is, it might make reasoning about TEI a bit difficult if we pick something that is not the actual id. Would be good to have @adrianplau input on which one is the best id

Reply:
Great! I missed that piece. My concern was exactly that, what is the idempotent identifier.

Reply:
Sorry, this one actually doesn't have a Sierra number attached to it yet! The xml id (manuscript_16172) comes from the Fihrist database, to which we also contribute our Arabic manuscript files. We're working on updating the Arabic files with Sierra numbers now and there might already be a batch of these coming later this week.

Reply:
Quick update: None of the Arabic files have the Sierra number entered yet, but we'll try to get this up relatively soon.

Reply:
Actually, I don't think it would be a good idea to use Sierra ids to store TEI because it would prevent us from storing files that don't have a sierra id. we will definitely need a sierra id for linking but that happens in another part of the pipeline. To store our replica of the TEI files, I think we should use whatever is the id in TEI world, which is why I went for `<idno type="msID">WMS_Arabic_529</idno>`. is that correct? Or would @jamesgorrie suggestion be more accurate?

Reply:
Sorry, then I understand! I think either is fine then. Following on James' suggestion, every TEI file does have a unique xml:id attribute to be valid, while the <idno type="msID"> tends to refer to the now canonical reference ID in the catalogue. In most cases, these overlap (though the Arabic ones have different xml:id attributes), but the xml:id attributes tend to be more consistent in format and are less likely to be edited. On the other hand, if we contribute to other catalogues, such as with the Arabic ones, there is the chance that there will be even more variation between <idno type="msID"> and <TEI xml:id="">. So perhaps stick with <idno type="msID">, as you suggest? Sorry again, this turned out to be a ramble!

Reply:
I guess necessary conditions for a field to be useful as an id are:
- It's unique across all TEI files. Ie no 2 files have the same value
- All files have exactly one value for that field. 
- It's not likely to change once created.
If the format varies across files it's unfortunate but less catastrophic than if we pick something that could be edited.
From what you're saying it looks like `xml:id` might be better as `<idno type="msID">` could be edited @adrianplau? or I'm not following 100%.

Reply:
Thank you, this is helpful. Yes, then I agree that <TEI xml:id=""> is better, as it ticks all three boxes. <idno type="msID"> could be edited if we were to update the cataloguing metadata with new references (this is what we see traces of in the <altIdentifier> in <msIdentifier>.

Reply:
Thanks @adrianplau, that is really useful. I've amended the document to reflect this